File: model_1/icpc-jakarta-2018/complete_1_solution.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    if (n == 0) {
        cout << 0 << endl;
        return 0;
    }

    vector<int> deg(n+1, 0);
    vector<pair<int, int>> edges;
    int total_edges = n * (n-1) / 2;

    for (int i=0; i<m; i++) {
        int a, b;
        cin >> a >> b;
        if (a > b) swap(a, b);
        edges.push_back({a, b});
        deg[a]++;
        deg[b]++;
    }

    if (m == total_edges) {
        cout << 0 << endl;
        return 0;
    }

    auto check = [&](int k) -> bool {
        vector<vector<char>> added(n+1, vector<char>(n+1, 0));
        vector<vector<char>> pushed(n+1, vector<char>(n+1, 0));
        vector<int> deg_local = deg;
        for (auto &e : edges) {
            added[e.first][e.second] = 1;
        }

        int added_count = m;
        queue<pair<int, int>> Q;

        for (int u=1; u<=n; u++) {
            for (int v=u+1; v<=n; v++) {
                if (added[u][v]) continue;
                if (deg_local[u] + deg_local[v] >= k) {
                    pushed[u][v] = 1;
                    Q.push({u, v});
                }
            }
        }

        while (!Q.empty()) {
            auto edge = Q.front();
            Q.pop();
            int u = edge.first;
            int v = edge.second;

            if (added[u][v]) 
                continue;

            added[u][v] = 1;
            added_count++;
            deg_local[u]++;
            deg_local[v]++;

            for (int w=1; w<=n; w++) {
                if (w == u || w == v) continue;
                int a = min(u, w);
                int b = max(u, w);
                if (!added[a][b] && !pushed[a][b]) {
                    if (deg_local[u] + deg_local[w] >= k) {
                        pushed[a][b] = 1;
                        Q.push({a, b});
                    }
                }
                int c = min(v, w);
                int d = max(v, w);
                if (!added[c][d] && !pushed[c][d]) {
                    if (deg_local[v] + deg_local[w] >= k) {
                        pushed[c][d] = 1;
                        Q.push({c, d});
                    }
                }
            }
        }

        return (added_count == total_edges);
    };

    int low = 0;
    int high = 2 * (n - 1);
    int ans = 0;

    while (low <= high) {
        int mid = (low + high) / 2;
        if (check(mid)) {
            ans = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    cout << ans << endl;
    return 0;
}

File: model_1/icpc-jakarta-2018/complete_1_solution.py
import sys
from collections import deque

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    n = int(data[0])
    m = int(data[1])
    graph = [[False] * (n+1) for _ in range(n+1)]
    deg = [0] * (n+1)
    idx = 2
    for _ in range(m):
        u = int(data[idx])
        v = int(data[idx+1])
        idx += 2
        graph[u][v] = True
        graph[v][u] = True
        deg[u] += 1
        deg[v] += 1

    total_edges = n * (n - 1) // 2
    target_edges_to_add = total_edges - m
    if target_edges_to_add == 0:
        print(0)
        return

    def check(k):
        if k == 0:
            return True
        g = [row[:] for row in graph]
        deg_copy = deg[:]
        q = deque()
        in_queue = set()
        
        for i in range(1, n+1):
            for j in range(i+1, n+1):
                if not g[i][j] and deg_copy[i] + deg_copy[j] >= k:
                    in_queue.add((i, j))
                    q.append((i, j))
                    
        count_added = 0
        while q and count_added < target_edges_to_add:
            i, j = q.popleft()
            in_queue.discard((i, j))
            
            if g[i][j]:
                continue
            if deg_copy[i] + deg_copy[j] < k:
                continue
                
            g[i][j] = True
            g[j][i] = True
            count_added += 1
            deg_copy[i] += 1
            deg_copy[j] += 1
            
            for node in (i, j):
                for neighbor in range(1, n+1):
                    if neighbor == node:
                        continue
                    if not g[node][neighbor]:
                        a, b = min(node, neighbor), max(node, neighbor)
                        if (a, b) not in in_queue:
                            s = deg_copy[node] + deg_copy[neighbor]
                            if s >= k:
                                in_queue.add((a, b))
                                q.append((a, b))
        return count_added == target_edges_to_add

    low = 0
    high = 2 * (n - 1)
    ans = 0
    while low <= high:
        mid = (low + high) // 2
        if check(mid):
            ans = mid
            low = mid + 1
        else:
            high = mid - 1
    print(ans)

if __name__ == '__main__':
    main()